### PR TITLE
Drop pre-3.0 GDAL support

### LINKED ISSUE #467: Does not compile with prebuilt GDAL 2.4 bindings
The gdal crate seems not to be compatible with the prebuilt bindings for GDAL 2.4 in `gdal_sys`, as some things missing from the bindings and not gated behind any feature flag in the `gdal` crate.
```
$ gdalinfo --version
GDAL 2.4.0, released 2018/12/14
```

Error snippet
```
error[E0425]: cannot find function, tuple struct or tuple variant `GDALGetGCPSpatialRef` in crate `gdal_sys`
   --> extlib/rust/gdal-0.16.0/src/gcp.rs:118:40
    |
118 |         let c_ptr = unsafe { gdal_sys::GDALGetGCPSpatialRef(self.c_dataset()) };
    |                                        ^^^^^^^^^^^^^^^^^^^^ not found in `gdal_sys`

error[E0425]: cannot find function, tuple struct or tuple variant `GDALSetGCPs2` in crate `gdal_sys`
   --> extlib/rust/gdal-0.16.0/src/gcp.rs:206:23
    |
206 |               gdal_sys::GDALSetGCPs2(
    |                         ^^^^^^^^^^^^ help: a function with a similar name exists: `GDALSetGCPs`
    |
    = note: similarly named function `GDALSetGCPs` defined here

error[E0425]: cannot find function, tuple struct or tuple variant `OCTNewCoordinateTransformationEx` in crate `gdal_sys`
  --> extlib/rust/gdal-0.16.0/src/spatial_ref/transform.rs:51:23
   |
51 |               gdal_sys::OCTNewCoordinateTransformationEx(
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `OCTNewCoordinateTransformation`
   |
   = note: similarly named function `OCTNewCoordinateTransformation` defined here

error[E0412]: cannot find type `OGRCoordinateTransformationOptionsH` in crate `gdal_sys`
  --> extlib/rust/gdal-0.16.0/src/spatial_ref/transform_opts.rs:14:22
   |
14 |     inner: gdal_sys::OGRCoordinateTransformationOptionsH,
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a type alias with a similar name exists: `OGRCoordinateTransformationH`
   |
   = note: similarly named type alias `OGRCoordinateTransformationH` defined here

error[E0425]: cannot find function, tuple struct or tuple variant `OCTDestroyCoordinateTransformationOptions` in crate `gdal_sys`
  --> extlib/rust/gdal-0.16.0/src/spatial_ref/transform_opts.rs:19:28
   |
19 |         unsafe { gdal_sys::OCTDestroyCoordinateTransformationOptions(self.inner) };
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `OCTDestroyCoordinateTransformation`
   |
   = note: similarly named function `OCTDestroyCoordinateTransformation` defined here

error[E0425]: cannot find function, tuple struct or tuple variant `OCTNewCoordinateTransformationOptions` in crate `gdal_sys`
  --> extlib/rust/gdal-0.16.0/src/spatial_ref/transform_opts.rs:26:40
   |
26 |           let c_obj = unsafe { gdal_sys::OCTNewCoordinateTransformationOptions() };
   |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: 
   |
   = note: similarly named function `OCTNewCoordinateTransformation` defined here

error[E0412]: cannot find type `OGRCoordinateTransformationOptionsH` in crate `gdal_sys`
  --> extlib/rust/gdal-0.16.0/src/spatial_ref/transform_opts.rs:39:56
   |
39 |     pub(crate) unsafe fn c_options(&self) -> gdal_sys::OGRCoordinateTransformationOptionsH {
   |                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a type alias with a similar name exists: `OGRCoordinateTransformationH`
   |
   = note: similarly named type alias `OGRCoordinateTransformationH` defined here

error[E0425]: cannot find function, tuple struct or tuple variant `OCTCoordinateTransformationOptionsSetAreaOfInterest` in crate `gdal_sys`
  --> extlib/rust/gdal-0.16.0/src/spatial_ref/transform_opts.rs:65:23
   |
65 |             gdal_sys::OCTCoordinateTransformationOptionsSetAreaOfInterest(
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `gdal_sys`

error[E0425]: cannot find function, tuple struct or tuple variant `OCTCoordinateTransformationOptionsSetOperation` in crate `gdal_sys`
   --> extlib/rust/gdal-0.16.0/src/spatial_ref/transform_opts.rs:145:23
    |
145 |             gdal_sys::OCTCoordinateTransformationOptionsSetOperation(
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `gdal_sys`

error[E0425]: cannot find function, tuple struct or tuple variant `OGR_G_MakeValid` in crate `gdal_sys`
   --> extlib/rust/gdal-0.16.0/src/vector/ops/transformations.rs:200:32
    |
200 |             unsafe { gdal_sys::OGR_G_MakeValid(self.c_geometry()) }
    |                                ^^^^^^^^^^^^^^^ help: a function with a similar name exists: `OGR_G_IsValid`
    |
    = note: similarly named function `OGR_G_IsValid` defined here
```

### LINKED ISSUE #477: Define a policy for the supported GDAL versions
We currently test on 3.1 to 3.8 in the official images, and on 3.0 on Ubuntu 20.04 (LTS), while 2.4 [is broken](https://github.com/georust/gdal/issues/467). We should decide if we want to fix the latter or drop the support, and in the long run we probably don't want to test dozens of versions.

Popular distributions have:
 - Debian 12 (stable): 3.6.2
 - Debian 11 (oldstable): 3.2.2
 - Ubuntu 22.04: 3.4.1
 - Ubuntu 20.04: 3.0.4
 - Ubuntu 18.04 (EOL 2023): 2.2.3
 - EL 9: 3.4.3
 - EL 8 (EOL 2029): 3.0.4
 - EL 7 (EOL 2024): 1.11.4

(you can check others on https://pkgs.org/, search for `gdal` and `gdal-bin`)

I'm inclined to say we should use Ubuntu LTS-1 (20.04) or Debian oldstable, however:

 - we're missing a lot of functionality, and I can see people wanting to get that while still supporting older systems
 - I'm not a fan of singling out Ubuntu in our policy

CC @rouault maybe you have some idea of which versions people are using.

### PR DESCRIPTION
**Pull request by lnicola**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/gdal/blob/master/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

Closes #477
Closes #467

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -14,28 +14,21 @@ fn main() {
...
-    // we only support GDAL >= 2.0.
-    for major in 2..=major {
+    for major in 3..=major {
         println!("cargo:rustc-cfg=major_ge_{major}");
     }
 
     for minor in 0..=minor {
         println!("cargo:rustc-cfg=minor_ge_{minor}");
     }
-
```
Comment (lnicola) [inline build.rs:37]: We never care about the patch version, so I think it's fine to emitting all those `cfg`s for it.

File: build.rs [hunk 2]
```diff
@@ -14,28 +14,21 @@ fn main() {
     let minor = (gdal_version - major * 1000000) / 10000;
     let patch = (gdal_version - major * 1000000 - minor * 10000) / 100;
```
Comment (rouault) [inline build.rs:15]: is that line calculating patch still needed ?

Comment (lnicola) [reply build.rs:15]: It's still printed in the panic message below, just in case. Probably doesn't hurt to have it.

Comment (rouault) [reply build.rs:15]: ah ok, nevermind, I missed that use

**Review by rouault — APPROVED** [review]

**Comment by lnicola** [comment]
Added version info support to the README and a commit to drop the pre-built 2.4 bindings. Will wait a little before merging, since it's the kind of change that might make people upset.

File: src/dataset.rs [hunk 3]
```diff
@@ -30,7 +30,6 @@ pub struct Dataset {
 // See: https://gdal.org/api/raster_c_api.html#_CPPv48GDALOpenPKc10GDALAccess
 // Additionally, VRT Datasets are not safe before GDAL 2.3.
 // See: https://gdal.org/drivers/raster/vrt.html#multi-threading-issues
```
Comment (metasim) [inline src/dataset.rs:32]: Is this comment still relevant?

**Review by metasim — APPROVED** [review]

**Comment by metasim** [comment]
@lnicola What do you think about holding back this PR until after the 0.17 release? Since the pre-3.0 support works up to now, it would give legacy users one more release (with lots of improvements) to work with.

**Comment by lnicola** [comment]
@metasim 2.4 is broken anyway, #467. We could fix it, but I don't think it's worth it since not many people complain.

**Comment by valarauca** [comment]
FYI the [`docs.rs` page for the project](https://docs.rs/gdal/0.16.0/gdal/) does state `2.4` to `3.X` is supported, so that should also be updated in this patch series.

**Comment by lnicola** [comment]
Thanks for spotting that, updated in https://github.com/georust/gdal/pull/504/commits/0f2996a3bed2b30393ff7de935ec2badbfcbd183.

**Comment by lnicola** [comment]
@metasim there wasn't much negative feedback, do you think we could merge this?

**Comment by metasim** [comment]
Yes! Go for it!

**Pull request merged**